### PR TITLE
fix: correctly open table when distributed datanode restart

### LINKED ISSUE #402: Cannot operate data after frontend restart
### Code version

Use branch `feat/distribute-poc`

### Examples

```
cargo run -- metasrv start --store-addr=127.0.0.1:2379

cargo run -- datanode start --rpc-addr=0.0.0.0:4100 --http-addr=0.0.0.0:4101 --mysql-addr=0.0.0.0:4102 --postgres-addr=0.0.0.0:4103 --node-id=1 
cargo run -- datanode start --rpc-addr=0.0.0.0:4200 --http-addr=0.0.0.0:4201 --mysql-addr=0.0.0.0:4202 --postgres-addr=0.0.0.0:4203 --node-id=2
cargo run -- datanode start --rpc-addr=0.0.0.0:4300 --http-addr=0.0.0.0:4301 --mysql-addr=0.0.0.0:4302 --postgres-addr=0.0.0.0:4303 --node-id=3

cargo run -- frontend start
```

Then use mysql client to create table:

```
mysql -h 127.0.0.1 -P 4002

CREATE TABLE dist_table(
    ts TIMESTAMP DEFAULT current_timestamp(),
    n INT,
    row_id INT,
    TIME INDEX (ts)
)
PARTITION BY RANGE COLUMNS (n) (
    PARTITION r0 VALUES LESS THAN (5),
    PARTITION r1 VALUES LESS THAN (9),
    PARTITION r2 VALUES LESS THAN (MAXVALUE),
)
engine=mito;

SELECT * FROM dist_table;
```

Now we restart frontend and use mysql client to reconnect:

```
mysql -h 127.0.0.1 -P 4002

mysql> select * from dist_table;
ERROR 2013 (HY000): Lost connection to MySQL server during query
No connection. Trying to reconnect...
Connection id:    8
Current database: *** NONE ***

ERROR 2013 (HY000): Lost connection to MySQL server during query
No connection. Trying to reconnect...
Connection id:    8
Current database: *** NONE ***

ERROR 2013 (HY000): Lost connection to MySQL server during query
mysql> INSERT INTO dist_table(n, row_id) VALUES (6, 6);
ERROR 2013 (HY000): Lost connection to MySQL server during query
No connection. Trying to reconnect...
Connection id:    8
Current database: *** NONE ***

ERROR 2013 (HY000): Lost connection to MySQL server during query
No connection. Trying to reconnect...
Connection id:    8
Current database: *** NONE ***

ERROR 2013 (HY000): Lost connection to MySQL server during query
```

### PR DESCRIPTION
**Pull request by MichaelScofield**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

fix #402 

Please explain IN DETAIL what the changes are in this PR and why they are needed:

- Summarize your change (**mandatory**)

1. correctly open table when distributed datanode restart
2. concat node_id in data dir and wal dir of datanode, when it is started in distributed mode
3. upgrade opensrv-mysql to latest version 0.2
4. add `--data-dir` and `--wal-dir` to command line options, to indicate dn's data dir and wal dir in distributed mode.

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

File: src/datanode/src/instance.rs [hunk 1]
```diff
@@ -169,21 +170,27 @@ impl Instance {
...
     // TODO(dennis): supports other backend
-    let data_dir = util::normalize_dir(match store_config {
+    let data_dir = match &opts.storage {
         ObjectStoreConfig::File { data_dir } => data_dir,
-    });
+    };
 
-    fs::create_dir_all(path::Path::new(&data_dir))
-        .context(error::CreateDirSnafu { dir: &data_dir })?;
+    let mut path = PathBuf::from(data_dir);
```
Comment (killme2008) [inline src/datanode/src/instance.rs:179]: It's too tricky, the user may be surprised that the data directory is not expected as they set.

I recommend to supports command line args `--wal-dir` and `--data-dir` , then we can modify the distributed getting start doc and pass these options by command line.

**Comment by MichaelScofield** [comment]
@killme2008 PTAL

File: src/frontend/src/instance/distributed.rs [hunk 2]
```diff
@@ -291,13 +307,13 @@ fn create_table_global_value(
 
     Ok(TableGlobalValue {
         node_id,
-        regions_id_map: HashMap::new(),
+        regions_id_map,
         table_info,
     })
 }
 
 // Remove this duplication in the future
```
Comment (killme2008) [inline src/frontend/src/instance/distributed.rs:315]: Adds a TODO?

File: src/mito/src/engine.rs [hunk 3]
```diff
@@ -404,7 +404,8 @@ impl<S: StorageEngine> MitoEngineInner<S> {
 
             let table_id = request.table_id;
             // TODO(dennis): supports multi regions;
-            let region_number = 0;
+            assert_eq!(request.region_numbers.len(), 1);
```
Comment (killme2008) [inline src/mito/src/engine.rs:407]:
```rust
assert_eq!(request.region_numbers, vec![0]);
```

File: src/servers/src/mysql/server.rs [hunk 4]
```diff
@@ -67,15 +72,28 @@ impl MysqlServer {
...
+        let spawn_result = io_runtime
+            .spawn(AsyncMysqlIntermediary::run_on(shim, r, w))
+            .await;
+        match spawn_result {
+            Ok(run_result) => {
+                if let Err(e) = run_result {
+                    error!(e; "Internal error occurred during query exec, server actively close the channel to let client try next time.")
+                }
+            }
+            Err(e) => error!("IO runtime cannot execute task, error: {}", e),
```
Comment (killme2008) [inline src/servers/src/mysql/server.rs:95]: In these two error cases, why we don't return an error result?

**Review by killme2008 — COMMENTED** [review]
Good job. The PR content may need to be edited too.

File: src/mito/src/engine.rs (hunk 3 above)
Comment (MichaelScofield) [reply src/mito/src/engine.rs:407]: `region_number` is assigned by Meta, might not necessarily be "0" here.

File: src/servers/src/mysql/server.rs (hunk 4 above)
Comment (MichaelScofield) [reply src/servers/src/mysql/server.rs:97]: Seems to write these 2 errors to client requires further investigation upon writing raw mysql protocol bytes into client channel. Allow me to add a todo here.

**Comment by MichaelScofield** [comment]
@killme2008 PTAL

**Review by killme2008 — APPROVED** [review]
LGTM

**Comment by MichaelScofield** [comment]
@v0y4g3r PTAL

**Review by v0y4g3r — APPROVED** [review]
Generally LGTM

File: src/servers/src/mysql/handler.rs [hunk 5]
```diff
@@ -145,6 +144,9 @@ impl<W: io::Write + Send + Sync> AsyncMysqlShim<W> for MysqlInstanceShim {
         query: &'a str,
         writer: QueryResultWriter<'a, W>,
     ) -> Result<()> {
+        debug!("Start executing query: '{}'", query);
```
[the remark this answers is not in the document]
Comment (MichaelScofield) [reply src/servers/src/mysql/handler.rs:147]: I've planned to use it later, when starting to design metrics.

**Pull request merged**